File: test/test.NatsUpstreamTransport.ts
import assert from "node:assert/strict";
import { spawn, type ChildProcess } from "node:child_process";
import { once } from "node:events";
import { writeFile, mkdtemp, rm } from "node:fs/promises";
import { tmpdir } from "node:os";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { getBinaryPath } from "@eplightning/nats-server";
import { test } from "vitest";

import { LocalBrowserLauncher } from "../bridge/LocalBrowserLauncher.js";
import { ModCDPClient } from "../client/js/ModCDPClient.js";

const HERE = path.dirname(fileURLToPath(import.meta.url));
const EXTENSION_PATH = path.resolve(HERE, "..", "dist", "extension");

function delay(ms: number) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function wait_for_websocket(url: string, timeout_ms = 10_000) {
  const deadline = Date.now() + timeout_ms;
  let last_error: unknown = null;
  while (Date.now() < deadline) {
    try {
      const ws = new WebSocket(url);
      await new Promise<void>((resolve, reject) => {
        ws.addEventListener("open", () => resolve(), { once: true });
        ws.addEventListener("error", reject, { once: true });
      });
      ws.close();
      return;
    } catch (error) {
      last_error = error;
      await delay(50);
    }
  }
  throw last_error instanceof Error ? last_error : new Error(`Timed out waiting for ${url}`);
}

async function close_process(proc: ChildProcess) {
  if (proc.exitCode != null || proc.signalCode != null) return;
  proc.kill("SIGTERM");
  await Promise.race([once(proc, "exit"), delay(2_000)]);
  if (proc.exitCode != null || proc.signalCode != null) return;
  proc.kill("SIGKILL");
  await Promise.race([once(proc, "exit"), delay(2_000)]);
}

async function start_nats_server() {
  const websocket_port = await LocalBrowserLauncher.freePort();
  const client_port = await LocalBrowserLauncher.freePort();
  const dir = await mkdtemp(path.join(tmpdir(), "modcdp-nats-"));
  const config_path = path.join(dir, "nats.conf");
  await writeFile(
    config_path,
    [
      `host: "127.0.0.1"`,
      `port: ${client_port}`,
      `websocket {`,
      `  host: "127.0.0.1"`,
      `  port: ${websocket_port}`,
      `  no_tls: true`,
      `}`,
      ``,
    ].join("\n"),
  );
  const proc = spawn(await getBinaryPath(), ["-c", config_path], { stdio: "ignore" });
  const url = `ws://127.0.0.1:${websocket_port}`;
  try {
    await wait_for_websocket(url);
  } catch (error) {
    await close_process(proc);
    await rm(dir, { recursive: true, force: true });
    throw error;
  }
  return {
    url,
    close: async () => {
      await close_process(proc);
      await rm(dir, { recursive: true, force: true });
    },
  };
}

test("nats upstream relays CDP through a real extension over a real NATS server", async () => {
  const nats = await start_nats_server();
  const subject_prefix = `modcdp.test.${Date.now()}`;
  const browser = new ModCDPClient({
    launch: {
      mode: "local",
      options: { headless: process.platform === "linux", sandbox: process.platform !== "linux" },
    },
    upstream: { mode: "ws" },
    extension: {
      mode: "auto",
      path: EXTENSION_PATH,
      service_worker_url_suffixes: ["/modcdp/service_worker.js"],
      trust_service_worker_target: true,
    },
  });
  let nats_client: ModCDPClient | null = null;

  try {
    await browser.connect();
    await browser.send("Mod.configure", {
      upstream: {
        mode: "nats",
        nats_url: nats.url,
        nats_subject_prefix: subject_prefix,
      },
      server: {
        loopback_cdp_url: browser.cdp_url,
        routes: { "*.*": "loopback_cdp" },
      },
    });

    nats_client = new ModCDPClient({
      launch: { mode: "none" },
      upstream: {
        mode: "nats",
        nats_url: nats.url,
        nats_subject_prefix: subject_prefix,
      },
      extension: { mode: "none" },
      server: {
        loopback_cdp_url: browser.cdp_url,
        routes: { "*.*": "loopback_cdp" },
      },
    });
    await nats_client.connect();
    assert.equal(nats_client.transport?.mode, "nats");
    assert.equal(nats_client.upstream_endpoint_kind, "modcdp_server");
    assert.equal(nats_client.transport?.url, `${nats.url}/`);
    assert.equal(nats_client.upstream.nats_subject_prefix, subject_prefix);
    const version = (await nats_client.send("Browser.getVersion")) as Record<string, unknown>;
    assert.equal(typeof version.product, "string");
  } finally {
    await nats_client?.close();
    await browser.close();
    await nats.close();
  }
}, 90_000);

File: client/python/tests/test_NatsUpstreamTransport.py
from __future__ import annotations

import socket
import subprocess
import tempfile
import threading
import time
import unittest
from pathlib import Path
from queue import Queue

from websocket import create_connection

from modcdp import ModCDPClient
from modcdp.NatsUpstreamTransport import NatsUpstreamTransport


ROOT = Path(__file__).resolve().parents[3]


class NatsUpstreamTransportTests(unittest.TestCase):
    def test_config_owns_url_subject_prefix_wait_timeout_and_injector_config(self) -> None:
        transport = NatsUpstreamTransport({"url": "ws://127.0.0.1:4223", "subject_prefix": "modcdp.one"})
        self.assertEqual(transport.url, "ws://127.0.0.1:4223/")
        self.assertEqual(transport.subject_prefix, "modcdp.one")
        self.assertEqual(
            transport.getInjectorConfig(),
            {"nats_url": "ws://127.0.0.1:4223/", "nats_subject_prefix": "modcdp.one"},
        )
        self.assertIs(
            transport.update(
                {
                    "nats_url": "nats://127.0.0.1:4222",
                    "nats_subject_prefix": "modcdp.two",
                    "role": "browser",
                    "wait_timeout_ms": 5,
                }
            ),
            transport,
        )
        self.assertEqual(transport.url, "nats://127.0.0.1:4222")
        self.assertEqual(transport.subject_prefix, "modcdp.two")
        with self.assertRaisesRegex(RuntimeError, "Timed out waiting 5ms for NATS ModCDP peer"):
            transport.waitForPeer()

    def test_close_rejects_pending_peer_waits(self) -> None:
        transport = NatsUpstreamTransport(
            {
                "url": "ws://127.0.0.1:4223",
                "subject_prefix": "modcdp.close",
                "wait_timeout_ms": 5_000,
            }
        )
        result: Queue[BaseException | None] = Queue()

        def wait_for_peer() -> None:
            try:
                transport.waitForPeer()
            except BaseException as error:
                result.put(error)
                return
            result.put(None)

        thread = threading.Thread(target=wait_for_peer, daemon=True)
        thread.start()
        time.sleep(0.05)
        transport.close()
        thread.join(timeout=1)

        error = result.get(timeout=1)
        self.assertIsInstance(error, RuntimeError)
        self.assertRegex(str(error), r"NATS transport for modcdp\.close closed before a peer connected")

    def test_close_resets_peer_wait_state(self) -> None:
        transport = NatsUpstreamTransport({"wait_timeout_ms": 5})
        transport._handle_payload('{"type":"modcdp.nats.hello","role":"browser","version":1}')

        transport.waitForPeer()
        transport.close()

        with self.assertRaisesRegex(RuntimeError, "Timed out waiting 5ms for NATS ModCDP peer"):
            transport.waitForPeer()
        self.assertTrue(transport.closed)

    def test_reconnect_after_close_resets_closed_state_with_real_nats_server(self) -> None:
        nats = _start_nats_server()
        transport = NatsUpstreamTransport({"url": nats["url"], "subject_prefix": f"modcdp.reconnect.{int(time.time() * 1000)}"})

        try:
            transport.connect()
            self.assertTrue(transport.connected)
            transport.close()
            self.assertFalse(transport.connected)
            self.assertTrue(transport.closed)
            transport.connect()
            self.assertTrue(transport.connected)
            self.assertFalse(transport.closed)
        finally:
            transport.close()
            nats["close"]()

    def test_relays_cdp_through_real_extension_over_real_nats_server(self) -> None:
        nats = _start_nats_server()
        subject_prefix = f"modcdp.test.{int(time.time() * 1000)}"
        cdp = ModCDPClient(
            launch={"mode": "local", "options": {"headless": True, "sandbox": False}},
            upstream={"mode": "nats", "nats_url": nats["url"], "nats_subject_prefix": subject_prefix},
            extension={
                "mode": "auto",
                "service_worker_url_suffixes": ["/modcdp/service_worker.js"],
                "trust_service_worker_target": True,
            },
            server={"routes": {"*.*": "loopback_cdp"}},
        )

        try:
            cdp.connect()
            self.assertEqual(cdp.transport.mode if cdp.transport else None, "nats")
            self.assertEqual(cdp.upstream_endpoint_kind, "modcdp_server")
            transport = cdp.transport
            if not isinstance(transport, NatsUpstreamTransport):
                self.fail(f"transport = {type(transport).__name__}")
            self.assertEqual(transport.url, f"{nats['url']}/")
            self.assertEqual(transport.subject_prefix, subject_prefix)
            version = cdp.send("Browser.getVersion")
            self.assertIsInstance(version["product"], str)
        finally:
            cdp.close()
            nats["close"]()


def _start_nats_server():
    websocket_port = _free_port()
    client_port = _free_port()
    temp_dir = tempfile.TemporaryDirectory(prefix="modcdp-nats-")
    config_path = Path(temp_dir.name) / "nats.conf"
    config_path.write_text(
        "\n".join(
            [
                'host: "127.0.0.1"',
                f"port: {client_port}",
                "websocket {",
                '  host: "127.0.0.1"',
                f"  port: {websocket_port}",
                "  no_tls: true",
                "}",
                "",
            ]
        )
    )
    binary_path = subprocess.check_output(
        [
            "pnpm",
            "exec",
            "node",
            "--input-type=module",
            "-e",
            "import { getBinaryPath } from '@eplightning/nats-server'; console.log(await getBinaryPath())",
        ],
        cwd=ROOT,
        text=True,
    ).strip()
    proc = subprocess.Popen([binary_path, "-c", str(config_path)], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
    url = f"ws://127.0.0.1:{websocket_port}"
    try:
        _wait_for_websocket(url)
    except Exception:
        _close_process(proc)
        temp_dir.cleanup()
        raise

    def close() -> None:
        _close_process(proc)
        temp_dir.cleanup()

    return {"url": url, "close": close}


def _wait_for_websocket(url: str, timeout_s: float = 10) -> None:
    deadline = time.time() + timeout_s
    last_error: Exception | None = None
    while time.time() < deadline:
        try:
            ws = create_connection(url, timeout=1)
            ws.close()
            return
        except Exception as error:
            last_error = error
            time.sleep(0.05)
    raise last_error or TimeoutError(f"Timed out waiting for {url}")


def _close_process(proc: subprocess.Popen) -> None:
    if proc.poll() is not None:
        return
    proc.terminate()
    try:
        proc.wait(timeout=2)
    except subprocess.TimeoutExpired:
        proc.kill()
        proc.wait(timeout=2)


def _free_port() -> int:
    sock = socket.socket()
    sock.bind(("127.0.0.1", 0))
    try:
        return int(sock.getsockname()[1])
    finally:
        sock.close()


if __name__ == "__main__":
    unittest.main()
